Heisenbug

Finally managed to get a stack trace from kafka when creating topics in the test suite, this is the cause of our occasionally randomly failing tests:

```
Error while executing topic command org.apache.zookeeper.KeeperException$NoNodeException: KeeperErrorCode = NoNode for /brokers/ids
org.I0Itec.zkclient.exception.ZkNoNodeException: org.apache.zookeeper.KeeperException$NoNodeException: KeeperErrorCode = NoNode for /brokers/ids
    at org.I0Itec.zkclient.exception.ZkException.create(ZkException.java:47)
    at org.I0Itec.zkclient.ZkClient.retryUntilConnected(ZkClient.java:685)
    at org.I0Itec.zkclient.ZkClient.getChildren(ZkClient.java:413)
    at org.I0Itec.zkclient.ZkClient.getChildren(ZkClient.java:409)
    at kafka.utils.ZkUtils$.getChildren(ZkUtils.scala:480)
    at kafka.utils.ZkUtils$.getSortedBrokerList(ZkUtils.scala:81)
    at kafka.admin.AdminUtils$.createTopic(AdminUtils.scala:154)
    at kafka.admin.TopicCommand$.createTopic(TopicCommand.scala:86)
    at kafka.admin.TopicCommand$.main(TopicCommand.scala:50)
    at kafka.admin.TopicCommand.main(TopicCommand.scala)
Caused by: org.apache.zookeeper.KeeperException$NoNodeException: KeeperErrorCode = NoNode for /brokers/ids
    at org.apache.zookeeper.KeeperException.create(KeeperException.java:102)
    at org.apache.zookeeper.KeeperException.create(KeeperException.java:42)
    at org.apache.zookeeper.ZooKeeper.getChildren(ZooKeeper.java:1249)
    at org.apache.zookeeper.ZooKeeper.getChildren(ZooKeeper.java:1277)
    at org.I0Itec.zkclient.ZkConnection.getChildren(ZkConnection.java:99)
    at org.I0Itec.zkclient.ZkClient$2.call(ZkClient.java:416)
    at org.I0Itec.zkclient.ZkClient$2.call(ZkClient.java:413)
    at org.I0Itec.zkclient.ZkClient.retryUntilConnected(ZkClient.java:675)
```

FWIW, this occurred on creating the first topic, while creating the second topic succeeded. I keep bumping the "wait for it to load" sleep call to no effect though, so I'm not sure what's going on.

--- a/vagrant/boot_cluster.sh
+++ b/vagrant/boot_cluster.sh
@@ -4,14 +4,15 @@ set -ex
 
 vagrant/run_toxiproxy.sh &
 
-sleep 5
+while ! nc -q 1 localhost 2181 </dev/null; do echo "Waiting"; sleep 1; done
+while ! nc -q 1 localhost 9092 </dev/null; do echo "Waiting"; sleep 1; done
 
 for i in 1 2 3 4 5; do
     KAFKA_PORT=`expr $i + 9090`
     cd ${KAFKA_INSTALL_ROOT}/kafka-${KAFKA_PORT} && bin/zookeeper-server-start.sh -daemon config/zookeeper.properties
 done
 
-sleep 5
+while ! nc -q 1 localhost 21805 </dev/null; do echo "Waiting"; sleep 1; done
 
 for i in 1 2 3 4 5; do
     KAFKA_PORT=`expr $i + 9090`

--- a/vagrant/create_topics.sh
+++ b/vagrant/create_topics.sh
@@ -2,7 +2,8 @@
 
 set -ex
 
-sleep 10
+while ! nc -q 1 localhost 21801 </dev/null; do echo "Waiting"; sleep 1; done
+while ! nc -q 1 localhost 29092 </dev/null; do echo "Waiting"; sleep 1; done
 
 cd ${KAFKA_INSTALL_ROOT}/kafka-9092
 bin/kafka-topics.sh --create --partitions 1 --replication-factor 3 --topic single_partition --zookeeper localhost:2181

--- a/vagrant/create_topics.sh
+++ b/vagrant/create_topics.sh
@@ -8,5 +8,3 @@ while ! nc -q 1 localhost 29092 </dev/null; do echo "Waiting"; sleep 1; done
 cd ${KAFKA_INSTALL_ROOT}/kafka-9092
 bin/kafka-topics.sh --create --partitions 1 --replication-factor 3 --topic single_partition --zookeeper localhost:2181
 bin/kafka-topics.sh --create --partitions 2 --replication-factor 3 --topic multi_partition  --zookeeper localhost:2181
-
-sleep 5

--- a/vagrant/run_toxiproxy.sh
+++ b/vagrant/run_toxiproxy.sh
@@ -5,7 +5,7 @@ set -ex
 ${KAFKA_INSTALL_ROOT}/toxiproxy -port 8474 -host 0.0.0.0 &
 PID=$!
 
-sleep 5
+while ! nc -q 1 localhost 8474 </dev/null; do echo "Waiting"; sleep 1; done
 
 wget -S --post-data='{"name":"zk1", "upstream":"localhost:21801", "listen":"0.0.0.0:2181"}' localhost:8474/proxies
 wget -S --post-data='{"name":"zk2", "upstream":"localhost:21802", "listen":"0.0.0.0:2182"}' localhost:8474/proxies

